### ISSUE TITLE
Transformer model memory mask to support Batching

### ISSUE THREAD

**Issue by philip30** [pytorch/pytorch#24461]
## 🚀 Feature
Currently the Transformer decoder (v1.2.0) module support (T, S) mask for the memory. It would be nice if we can support (T, S, N) so that we can mask different source for different batch. 

## Motivation

This is good where different masking of source is requested for different target batch. Use case: simultaneous translation system.

## Alternatives

Currently the only work around is to batch examples who have same memory masks or loop the example one by one.

cc @SsnL

**Comment by ZhuBaohe** [comment]
You can try the parameters ‘src_key_padding_mask’ and ‘tgt_key_padding_mask’

**Comment by philip30** [comment]
I guess they are a bit different. Here the complexity is that we want to mask sum number of source, for each target, for each batch. Can you show me how to do it with src_key_padding_mask and tgt_key_padding_mask?

**Comment by ZhuBaohe** [comment]
@philip30  I correct my statement，you can try the parameters ‘memory_key_padding_mask’ whose size  is (N, S), where S is the source sequence length, N is the batch size.

**Comment by philip30** [comment]
@ZhuBaohe sorry, this is still not answering my second explanation. The complexity should be TXSXN, and can't be modeled otherwise I guess.

**Comment by ZhuBaohe** [comment]
I agree you. I think these parameters, i.e.,  'memory_mask' and 'memory_key_padding_mask' , should be merged one mask parameter whose size is (N, T, S) . It's useful for some translation tasks.

**Comment by ZheruiHuang** [comment]
Hello, sorry for boarding you in this long time ago issue. I read pytorch's docs (v2.3) and noticed that `nn.MultiheadAttention` had supported 3D attn_mask, but `nn.TransformerDecoder` seems to not support 3D memory_mask yet (see https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html#torch.nn.Transformer). Have you solved your problem in another way instead of calling `nn.TransformerDecoder.forward`?
